pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -72,21 +76,36 @@ pub async fn download_layer_file<'a>(
         layer_metadata.generation,
     );
 
-    // Perform a rename inspired by durable_rename from file_utils.c.
-    // The sequence:
-    //     write(tmp)
-    //     fsync(tmp)
-    //     rename(tmp, new)
-    //     fsync(new)
-    //     fsync(parent)
-    // For more context about durable_rename check this email from postgres mailing list:
-    // https://www.postgresql.org/message-id/56583BDD.9060302@2ndquadrant.com
-    // If pageserver crashes the temp file will be deleted on startup and re-downloaded.
-    let temp_file_path = path_with_suffix_extension(local_path, TEMP_DOWNLOAD_EXTENSION);
-
-    let bytes_amount = download_retry(
+    let (bytes_amount, temp_file) = download_retry(
         || async {
-            download_object(storage, &remote_path, &temp_file_path, gate, cancel, ctx).await
+            // TempVirtualFile requires us to never reuse a filename while an old
+            // instance of TempVirtualFile created with that filename is not done dropping yet.
+            // So, we use a monotonic counter to disambiguate the filenames.
+            static NEXT_TEMP_DISAMBIGUATOR: AtomicU64 = AtomicU64::new(1);
+            let filename_disambiguator =
+                NEXT_TEMP_DISAMBIGUATOR.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+
+            let temp_file_path = path_with_suffix_extension(
+                local_path,
+                &format!("{filename_disambiguator:x}.{TEMP_DOWNLOAD_EXTENSION}"),
+            );
+
+            let temp_file = TempVirtualFile::new(
+                // Not _v2 yet which is sensitive to virtual_file_io_mode.
+                // That'll happen in PR https://github.com/neondatabase/neon/pull/11558
+                VirtualFile::open_with_options(
+                    &temp_file_path,
+                    virtual_file::OpenOptions::new()
+                        .create_new(true)
+                        .write(true),
+                    ctx,
+                )
+                .await
```

Comment:
This assumes that any stray files will be removed on process restart, otherwise the `create_new()` below will error out. That pretty brittle considering we restart from 1. I don't know if we can make that assumption currently, but afaik the Pageserver will remove unknown files on startup.

I think I'd prefer to just bite the bullet and harden `TempVirtualFile` to properly manage the path and file right away, rather than leave it in this intermediate limbo. At best, it's a footgun.

Reply:
> This assumes that any stray files will be removed on process restart, otherwise the create_new() below will error out. That pretty brittle considering we restart from 1. I don't know if we can make that assumption currently, but afaik the Pageserver will remove unknown files on startup.

We can rely on this, and have for many years.

Reply:
I agree that TempVirtualFile should manage it, but, let's get the perf-sensitive patch series in first (this PR + https://github.com/neondatabase/neon/pull/11558 ) and then tend to centralizing the logic.